apps/wallet/src/background/storage-migration.ts

```diff
@@ -0,0 +1,150 @@
+		);
+	}
+	return { mnemonicSource, mnemonicAccounts };
+}
+
+async function makeImportedAccounts(password: string) {
+	const importedKeyPairs = VaultStorage.getImportedKeys();
+	if (!importedKeyPairs) {
+		throw new Error('Failed to load imported accounts, vault is locked');
+	}
+	return Promise.all(
+		importedKeyPairs.map((keyPair) =>
+			ImportedAccount.createNew({ password, keyPair: keyPair.export() }),
+		),
+	);
+}
+
+async function makeLedgerAccounts(password: string) {
+	const ledgerAccounts = await getSavedLedgerAccounts();
+	return Promise.all(
+		ledgerAccounts.map(({ address, derivationPath, publicKey }) =>
+			LedgerAccount.createNew({ address, derivationPath, password, publicKey }),
+		),
+	);
+}
+
+async function makeQredoAccounts(password: string) {
+	const qredoSources = [];
+	const qredoAccounts: Omit<QredoSerializedAccount, 'id'>[] = [];
+	for (const aQredoConnection of await getAllQredoConnections()) {
+		const refreshToken = VaultStorage.getQredoToken(aQredoConnection.id);
+		if (!refreshToken) {
+			throw new Error(
+				`Failed to load qredo account (${aQredoConnection.id}), refresh token not found`,
+			);
+		}
+		const aQredoSource = await QredoAccountSource.createNew({
+			password,
+			apiUrl: aQredoConnection.apiUrl,
+			organization: aQredoConnection.origin,
+			origin: aQredoConnection.origin,
+			service: aQredoConnection.service,
+			refreshToken,
+		});
+		qredoSources.push(aQredoSource);
+		for (const aWallet of aQredoConnection.accounts) {
+			qredoAccounts.push({
+				...aWallet,
+				type: 'qredo',
+				lastUnlockedOn: null,
+				sourceID: aQredoSource.id,
+			});
+		}
+	}
+	return { qredoSources, qredoAccounts };
+}
+
+export async function doMigration(password: string) {
+	await VaultStorage.unlock(password);
+	const currentStatus = await getStatus();
+	if (currentStatus === 'required') {
+		statusCache = 'inProgress';
+		try {
+			const db = await getDB();
+			const { mnemonicAccounts, mnemonicSource } = await makeMnemonicAccounts(password);
+			const importedAccounts = await makeImportedAccounts(password);
+			const ledgerAccounts = await makeLedgerAccounts(password);
+			const { qredoAccounts, qredoSources } = await makeQredoAccounts(password);
+			await db.transaction('rw', db.accounts, db.accountSources, async () => {
+				await MnemonicAccountSource.save(mnemonicSource, { skipBackup: true });
+				await addNewAccounts(mnemonicAccounts, { skipBackup: true });
+				await addNewAccounts(importedAccounts, { skipBackup: true });
+				await addNewAccounts(ledgerAccounts, { skipBackup: true });
+				for (const aQredoSource of qredoSources) {
+					await QredoAccountSource.save(aQredoSource, { skipBackup: true });
+				}
+				await addNewAccounts(qredoAccounts, { skipBackup: true });
+				await Dexie.waitFor(setToLocalStorage(migrationDoneStorageKey, true));
+			});
+			statusCache = 'ready';
+			backupDB();
+		} catch (e) {
+			statusCache = 'required';
```

Comment:
I hope that it will be a recoverable error and when user tries again it will succeed. If migration doesn't work then the app will be unusable anyway.
We can add a way to init/clear storage and start fresh in order to hide it